### PR TITLE
Added backend logging using Log4js

### LINKED ISSUE #114: Documentation
Since the Fractal web client is approaching its first usable version, having some documentation will soon become important. We briefly discussed with @mfranzon about it. Instead of focusing on the choice of a specific framework, here are a few high-level points that we see as a good start.

1. Target readers:
    * The Fractal dev team, in 6 months or one year: we will obviously not remember why a certain choice was made.
    * Other research groups who want to give a try to Fractal. If they go above a minimal commitment threshold, and setup a fractal-server instance, they may as well just spin up a web client to make its use simpler. Without some basic docs, this is not a real option.
2. We should not aim at autogenerated documentation for all components/functions/modules of the package (like the one on fractal-server or fractal-tasks-core). Still, we should make sure that code has a decent amount of explanatory comments (which we could, in the future, turn into standardize docstrings to be parsed with some tool).
3. As a first version, let's aim for the least possible effort in maintaining/serving the docs. If there is no ready-to-use framework, we can even start with a `docs` folder containing a few markdown files - and we'll later set up a proper documentation (compiled and served, e.g. on github pages).
4. What should be there:
    * How to install the package (possibly with common gotchas).
    * How to run the package (possibly with common gotchas).
    * Known limitations. Examples are: (A) safari not (currently) supported, (B) a certain fractal-server feature/version is not supported, ...
    * Release notes (this means that we'll start having some kind of release, even just on github, and to keep an up-to-date CHANGELOG.md file in the repository root)
    * Description of specific technical choices. Random examples: (A) the use of a svelte store for the project context, (B) the choice of `datatables` to show the jobs table, (C) how we (will) implement the use of JSON schemas for task arguments, ... This should be a more structured version of a development log, which we can look back at in the future when we need to review and possibly override a certain choice.

Given the experience with the other Fractal repositories, I'd put the stress on point 3: let's not get lost in the huge landscape of documentation frameworks. If we have time to invest, let's use it to write some clear and complete markdown files - and then we'll find a decent way to publish them..

### LINKED ISSUE #487: Remove log from `handleFetch`
Long time ago I added the following log:
```
Including cookie into request to http://..., via handleFetch
```
(due to https://github.com/fractal-analytics-platform/fractal-web/issues/274, and fixed with https://github.com/fractal-analytics-platform/fractal-web/pull/275/files)

This log line is very frequent, but not very informative. Let's either place it at a "debug" logging level (if this works like in Python), or simply remove it.

### LINKED ISSUE #489: Review server logging (possibly via third-party library)
`console.log` is fine, but AFAIK it doesn't offer easy format customization, logging levels, or redirection to file. Let's see whether we find an appropriate 3rd-party library.

The basic criteria are something like
1. Lightweight (in terms of dependencies and configuration effort).
2. Widespread and maintained (and not necessarily a svelte-specific one)
3. Required feature: log record format (e.g. we want to include timestamps)
4. Required feature: logging levels
5. Useful feature: logging to file
6. Anything else?

I don't have an opinion (or even a clear understanding) on whether the same logging strategy should be adopted for browser logging as well.

### LINKED ISSUE #494: Fix admin-area download-job-logs endpoint path
> (I did not investigate why the backend is sending 404, for the moment)

The endpoint is probably lacking a `v2`. It should be `/admin/v2/job/{job_id}/download/`.

_Originally posted by @tcompa in https://github.com/fractal-analytics-platform/fractal-web/issues/493#issuecomment-2126679339_

### LINKED ISSUE #495: Document which env variables are static/dynamic
When running with `node` (after a build), the dynamic variables should be provided as part of the environment - let's make sure we know which ones they are.

### LINKED ISSUE #499: Updating workflow tasks to new version fails in Fractal legacy
I can no longer update workflow tasks to a newer version in Fractal legacy. Can we have a look on what may have broken there?

I get errors like:
Screenshot 2024-05-28 at 14 34 47

Sometimes it says Project not found, sometimes it says not allowed for project 2.

Does it maybe look in the V2 projects?

### PR DESCRIPTION
**Pull request by zonia3000**
## Checklist before merging
- [x] I added an appropriate entry to `CHANGELOG.md`

Closes #489
Closes #487
Closes #495
Closes #494
Closes #499
Close #114

Logic used to set the log level of the added/modified log messages:

* `INFO` level for each request detected in `hooks.server.js`, since they display the URL of all the HTTP calls that reach the Svelte backend;
* `DEBUG` for messages describing that we are going to perform a specific call to fractal-server (e.g. "Fetching the list of projects");
* `ERROR` for error responses in endpoints that should always return a success response (e.g. the projects list);
* `INFO` for errors responses where the failure is a valid state (e.g. unsuccessful login);
* `TRACE` for the internal proxy, since the messages are usually a duplication of the one present in the hooks component, unless there is something wrong with the proxy itself;
* `TRACE` for messaging describing that SSR of a given page is happening (e.g. "Loading jobs page"), since we can usually understand this also from the messages generated in the hooks component.

I've also done a bit of refactoring of the methods in Svelte backend in order to improve typing and I've removed some of methods that weren't used.

### CODE REVIEW

**Comment by tcompa** [comment]
Minor comment: could we change the timestamp format from `2024-05-23T09:20:20.893` to `2024-05-23 09:20:12,745`? This way it would be the same as for the backend logs, which is useful when showing them together.

Here is an example from fractal-containers:
```
fractal-web       | [2024-05-23T09:22:04.120] [INFO] (hooks) - [GET] - /v2/tasks
fractal-server    | 2024-05-23 09:22:04,125 - uvicorn.access - INFO - 172.19.0.4:49870 - "GET /auth/current-user/ HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:04,136 - uvicorn.access - INFO - 172.19.0.4:49886 - "GET /api/v2/task/?args_schema_non_parallel=false&args_schema_parallel=false HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:04,245 - uvicorn.access - INFO - 127.0.0.1:56614 - "GET /api/alive/ HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:04,574 - uvicorn.access - INFO - 172.19.0.5:58294 - "GET /api/v2/project/1/job/1/ HTTP/1.1" 200
fractal-demos-v2  | [v2] examples/01   "status": "submitted",
fractal-server    | 2024-05-23 09:22:05,192 - uvicorn.access - INFO - 172.19.0.6:40912 - "GET /api/v1/project/1/job/1/ HTTP/1.1" 200
fractal-web       | [2024-05-23T09:22:06.077] [INFO] (hooks) - [GET] - /v2/projects
fractal-server    | 2024-05-23 09:22:06,083 - uvicorn.access - INFO - 172.19.0.4:49870 - "GET /auth/current-user/ HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:06,094 - uvicorn.access - INFO - 172.19.0.4:49886 - "GET /api/v2/project/ HTTP/1.1" 200
fractal-web       | [2024-05-23T09:22:07.290] [INFO] (hooks) - [GET] - /v2/projects/1
fractal-server    | 2024-05-23 09:22:07,294 - uvicorn.access - INFO - 172.19.0.4:49870 - "GET /auth/current-user/ HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:07,300 - uvicorn.access - INFO - 172.19.0.4:49886 - "GET /api/v2/project/1/ HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:07,312 - uvicorn.access - INFO - 172.19.0.4:49870 - "GET /api/v2/project/1/dataset/?history=false HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:07,331 - uvicorn.access - INFO - 172.19.0.4:49886 - "GET /api/v2/project/1/workflow/ HTTP/1.1" 200
fractal-server    | 2024-05-23 09:22:08,759 - uvicorn.access - INFO - 172.19.0.5:52156 - "GET /api/v2/project/1/job/1/ HTTP/1.1" 200
```

---

Having them all homogeneous means that we can easily combine logfiles from different sources and sort record.

**Comment by tcompa** [comment]
Quick question about the log file:

1. It is created with `-rw-------` permissions - is this on purpose? If so, no problem with that.

**Comment by tcompa** [comment]
> It is created with `-rw-------` permissions - is this on purpose? If so, no problem with that.

Answer: https://github.com/log4js-node/log4js-node/security/advisories/GHSA-82v2-mx6x-wq7q
(thanks @zonia3000)

**Comment by zonia3000** [comment]
> Quick question about the three new env variables:

> 1. Are they read during the build phase, or only when running?

They are read at runtime.

> 2. Are they required, or do they have defaults in case they are not present in the env file?

They are not required. Default values are `warn` for `LOG_LEVEL_CONSOLE` and `info` for `LOG_LEVEL_FILE`.
If `LOG_FILE` is not present no file will be created.

**Review by tcompa — APPROVED** [review]
Nothing to add here, thanks!

(feel free to include https://github.com/fractal-analytics-platform/fractal-web/issues/495 here or in another PR)

File: docs/environment-variables.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Environment variables
+
+The following environment variables can be used to configure fractal-web. 
+
+> Only when using `npm run dev` or `npm run preview` these variables are automatically read from `.env.development` and `.env` files. In production they must be set in the shell that starts the node process.
+
+* `FRACTAL_SERVER_HOST`: represents the URL of the fractal-server application (e.g. http://localhost:8000, or http://subdomain.example.org:8000); this variable is **required** and it should not contain a final slash;
```
Comment (tcompa) [inline docs/environment-variables.md:7]: let's remove the trailing slash, if any

File: docs/quickstart.md [hunk 2]
```diff
@@ -44,17 +44,16 @@ You can create a script with the following content to run fractal-web installed
 export FRACTAL_SERVER_HOST=http://localhost:8000
 export PUBLIC_FRACTAL_ADMIN_SUPPORT_EMAIL=help@localhost
 export PUBLIC_OAUTH_CLIENT_NAME=
-# remember to set this in production
+# remember to set this in production (e.g. subdomain.example.org)
 export AUTH_COOKIE_DOMAIN=
-# set this to true in production
-export AUTH_COOKIE_SECURE=false
```
Comment (tcompa) [inline docs/quickstart.md]: to verify

Comment (zonia3000) [reply docs/quickstart.md]: I confirm that `Secure=true` works on localhost.

Comment (tcompa) [reply docs/quickstart.md]: Great, thanks. Then let's not include it, in order to keep it brief

**Pull request merged**